{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "fb9046e5-bdda-4924-94fb-6e27c3f81a85",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#!pip install biopython\n",
    "# Requirements: biopython\n",
    "# https://agat.readthedocs.io/en/latest/index.html\n",
    "# https://github.com/NBISweden/AGAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "43180866-ea0b-4e74-a7ce-40cd25b2acbe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d82b124e-3cef-4cb4-b350-da482bd10170",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import subprocess\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "45c37746-858f-4482-bf80-00751ce80d79",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Using standard /opt/anaconda3/envs/py39/lib/perl5/site_perl/auto/share/dist/AGAT/agat_config.yaml file\n",
      "GTF version relax selected from the agat config file.\n",
      "Reading input file\n",
      "\n",
      " ------------------------------------------------------------------------------\n",
      "|   Another GFF Analysis Toolkit (AGAT) - Version: v1.2.0                      |\n",
      "|   https://github.com/NBISweden/AGAT                                          |\n",
      "|   National Bioinformatics Infrastructure Sweden (NBIS) - www.nbis.se         |\n",
      " ------------------------------------------------------------------------------\n",
      "                                        \n",
      "                                       \n",
      "                          ------ Start parsing ------                           \n",
      "-------------------------- parse options and metadata --------------------------\n",
      "=> Accessing the feature_levels YAML file\n",
      "Using standard /opt/anaconda3/envs/py39/lib/perl5/site_perl/auto/share/dist/AGAT/feature_levels.yaml file\n",
      "=> Attribute used to group features when no Parent/ID relationship exists (i.e common tag):\n",
      "\t* locus_tag\n",
      "\t* gene_id\n",
      "=> merge_loci option deactivated\n",
      "=> Machine information:\n",
      "\tThis script is being run by perl v5.32.1\n",
      "\tBioperl location being used: /opt/anaconda3/envs/py39/lib/perl5/site_perl/Bio/\n",
      "\tOperating system being used: darwin \n",
      "=> Accessing Ontology\n",
      "\tfeature-ontology URI defined within the file: so.obo\n",
      "400 URL must be absolute\tThe URI provided (so.obo) doesn't work.\n",
      "error: Died at /opt/anaconda3/envs/py39/lib/perl5/site_perl/AGAT/OmniscientI.pm line 3772.\n",
      "\n",
      "\tNo ontology accessible from the gff file header!\n",
      "\tWe use the SOFA ontology distributed with AGAT:\n",
      "\t\t/opt/anaconda3/envs/py39/lib/perl5/site_perl/auto/share/dist/AGAT/so.obo\n",
      "\tRead ontology /opt/anaconda3/envs/py39/lib/perl5/site_perl/auto/share/dist/AGAT/so.obo:\n",
      "\t\t4 root terms, and 2596 total terms, and 1516 leaf terms\n",
      "\tFiltering ontology:\n",
      "\t\tWe found 1861 terms that are sequence_feature or is_a child of it.\n",
      "--------------------------------- parsing file ---------------------------------\n",
      "=> Number of line in file: 913929\n",
      "=> Number of comment lines: 11\n",
      "=> Fasta included: No\n",
      "=> Number of features lines: 76701\n",
      "=> Number of feature type (3rd column): 6\n",
      "\t* Level1: 1 => gene\n",
      "\t* level2: 2 => mRNA tRNA\n",
      "\t* level3: 2 => CDS exon\n",
      "\t* unknown: 1 => supercontig\n",
      "=> Version of the Bioperl GFF parser selected by AGAT: 3\n",
      "                                                                                \n",
      "--------------------- WARNING ---------------------\n",
      "MSG: ##feature-ontology header tag parsing unimplemented\n",
      "---------------------------------------------------\n",
      "                                                                                \n",
      "--------------------- WARNING ---------------------\n",
      "MSG: ##attribute-ontology header tag parsing unimplemented\n",
      "---------------------------------------------------\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "gff3 reader warning: primary_tag error @ supercontig still not taken into account! Please modify the feature_levels YAML file to define the feature in one of the levels.\n",
      "1 warning messages: \n",
      "--------------------- WARNING ---------------------\n",
      "MSG: ##attribute-ontology header tag parsing unimplemented\n",
      "---------------------------------------------------\n",
      "\n",
      "8 warning messages: gff3 reader warning: primary_tag error \n",
      "1 warning messages: \n",
      "--------------------- WARNING ---------------------\n",
      "MSG: ##feature-ontology header tag parsing unimplemented\n",
      "---------------------------------------------------\n",
      "\n",
      "                  ------ End parsing (done in 25 second) ------                 \n",
      "\n",
      "\n",
      "                           ------ Start checks ------                           \n",
      "---------------------------- Check1: feature types -----------------------------\n",
      "----------------------------------- ontology -----------------------------------\n",
      "All feature types in agreement with the Ontology.\n",
      "------------------------------------- agat -------------------------------------\n",
      "WARNING - Feature types not expected by AGAT:\n",
      "* supercontig\n",
      "The feature of these types (3rd column in GFF3) are skipped by the parser!\n",
      "To take them into account you must update the feature_levels YAML file. To acces\n",
      "s this file run:\n",
      "\t\t\tagat levels --expose\n",
      "In which file to add my feature?\n",
      "* Feature level1 (e.g. gene, match, region):\n",
      "  My feature has no parent\n",
      "  => level1 section.\n",
      "* Feature level2 (e.g. mrna, match_part, trna):\n",
      "  My feature has one parent and children\n",
      "  => level2 section.\n",
      "* Feature level3 (e.g. exon, intron, cds):\n",
      "  My feature has one parent (the parent has also a parent) and no children\n",
      "  => level3 section.\n",
      "* Feature level3 discontinuous (e.g. cds, utr):\n",
      "  A single feature that exists over multiple genomic locations\n",
      "  => spread section.\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "------------------------------ Check2: duplicates ------------------------------\n",
      "None found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "-------------------------- Check3: sequential bucket ---------------------------\n",
      "None found\n",
      "------------------------------ done in 2 seconds -------------------------------\n",
      "\n",
      "--------------------------- Check4: l2 linked to l3 ----------------------------\n",
      "No problem found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "--------------------------- Check5: l1 linked to l2 ----------------------------\n",
      "No problem found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "--------------------------- Check6: remove orphan l1 ---------------------------\n",
      "We remove only those not supposed to be orphan\n",
      "None found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "------------------------- Check7: all level3 locations -------------------------\n",
      "------------------------------ done in 2 seconds -------------------------------\n",
      "\n",
      "------------------------------ Check8: check cds -------------------------------\n",
      "No problem found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "\n",
      "----------------------------- Check9: check exons ------------------------------\n",
      "No exons created\n",
      "No exons locations modified\n",
      "No supernumerary exons removed\n",
      "No level2 locations modified\n",
      "------------------------------ done in 2 seconds -------------------------------\n",
      "\n",
      "----------------------------- Check10: check utrs ------------------------------\n",
      "4 UTRs created that were missing\n",
      "No UTRs locations modified\n",
      "No supernumerary UTRs removed\n",
      "------------------------------ done in 2 seconds -------------------------------\n",
      "\n",
      "------------------------ Check11: all level2 locations -------------------------\n",
      "No problem found\n",
      "------------------------------ done in 2 seconds -------------------------------\n",
      "\n",
      "------------------------ Check12: all level1 locations -------------------------\n",
      "No problem found\n",
      "------------------------------ done in 1 seconds -------------------------------\n",
      "\n",
      "---------------------- Check13: remove identical isoforms ----------------------\n",
      "None found\n",
      "------------------------------ done in 0 seconds -------------------------------\n",
      "                  ------ End checks (done in 11 second) ------                  \n",
      "\n",
      "\n",
      "converting to GTFrelax\n",
      "Formating output to GTFrelax\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=['agat_convert_sp_gff2gtf.pl', '--gff', '../dl_data/a_fumigatus_Af293.gff', '--out', '../data/a_fumigatus_Af293.gtf2'], returncode=0)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace 'input.gff' and 'output.gtf' with your actual file paths\n",
    "input_gff = '../dl_data/a_fumigatus_Af293.gff'\n",
    "output_gtf = '../data/a_fumigatus_Af293.gtf'\n",
    "\n",
    "# Build the command to call agat_convert_sp_gxf2gxf\n",
    "# agat_convert_sp_gff2gtf.pl --gff 1_test.gff -o 1_test_agat.gtf\n",
    "agat_command = [\n",
    "    'agat_convert_sp_gff2gtf.pl',\n",
    "    '--gff', input_gff,\n",
    "    '--out', output_gtf\n",
    "]\n",
    "\n",
    "# Run agat_convert_sp_gxf2gxf.py as a subprocess\n",
    "subprocess.run(agat_command, check=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a77d9120-8df4-4805-a687-73358433ffc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "fasta = open('../data/Afumigatus_AF293B_Genome.fasta', 'r').read()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da394800-b747-4bf1-a790-4bee13466fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "re.findall(r'>[A-Z]+', fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e287a44-1a2c-4b7e-8cb5-bfbde9405e91",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59d5e049-a1e5-46fc-adf5-04204e93fdf9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3b76e76-5bf8-4213-a6bd-0f333bebe32f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
